lib/active_admin/router.rb

```diff
@@ -1,108 +1,113 @@
-    #     ActiveAdmin.routes(self)
-    #   end
-    #
-    def apply(router)
-      define_root_routes router
-      define_resource_routes router
+    def apply
+      define_root_routes
+      define_resources_routes
     end
 
-    def define_root_routes(router)
-      router.instance_exec @application.namespaces do |namespaces|
-        namespaces.each do |namespace|
-          if namespace.root?
-            root namespace.root_to_options.merge(to: namespace.root_to)
-          else
-            namespace namespace.name, namespace.route_options.dup do
-              root namespace.root_to_options.merge(to: namespace.root_to, as: :root)
-            end
+    private
+
+    def define_root_routes
+      namespaces.each do |namespace|
+        if namespace.root?
+          router.root namespace.root_to_options.merge(to: namespace.root_to)
+        else
+          router.namespace namespace.name, namespace.route_options.dup do
+            router.root namespace.root_to_options.merge(to: namespace.root_to, as: :root)
           end
         end
       end
     end
 
     # Defines the routes for each resource
-    def define_resource_routes(router)
-      router.instance_exec @application.namespaces, self do |namespaces, aa_router|
-        resources = namespaces.flat_map{ |n| n.resources.values }
-        resources.each do |config|
-          routes = aa_router.resource_routes(config)
-
-          # Add in the parent if it exists
-          if config.belongs_to?
-            belongs_to = routes
-            routes     = Proc.new do
-              # If it's optional, make the normal resource routes
-              instance_exec &belongs_to if config.belongs_to_config.optional?
-
-              # Make the nested belongs_to routes
-              # :only is set to nothing so that we don't clobber any existing routes on the resource
-              resources config.belongs_to_config.target.resource_name.plural, only: [] do
-                instance_exec &belongs_to
-              end
-            end
-          end
+    def define_resources_routes
+      resources = namespaces.flat_map{ |n| n.resources.values }
+      resources.each do |config|
+        define_resource_routes(config)
+      end
+    end
 
-          # Add on the namespace if required
-          unless config.namespace.root?
-            nested = routes
-            routes = Proc.new do
-              namespace config.namespace.name, config.namespace.route_options.dup do
-                instance_exec &nested
-              end
-            end
-          end
+    def define_resource_routes(config)
+      if config.namespace.root?
+        define_routes(config)
+      else
+        # Add on the namespace if required
+        define_namespace(config)
+      end
+    end
 
-          instance_exec &routes
-        end
+    def define_routes(config)
+      if config.belongs_to?
+        define_belongs_to_routes(config)
+      else
+        page_or_resource_routes(config)
```

Comment:
For consistency, should these methods be named either `define_belongs_to_routes` & `define_page_or_resource_routes`, or `belongs_to_routes` & `page_or_resource_routes`?

Reply:
I agree the method naming is not great, I was focusing on unfolding the code and followed the examples of #3170.  Renaming methods retroactively with an interactive rebase can get tricky so I'll look at a final commit to do this.

Reply:
On second thoughts reviewing method names just left me wanting to refactor even further. I'ld rather close this technical excision of proc/exec and leave further improvements to a separate PR.